### DISCUSSION TITLE
Pluralization question

### DISCUSSION THREAD

**Discussion by ItsRobbAllen** [ravendb/ravendb#14584] [General]
How does Raven pluralize the object name and is that something that is available somewhere?

There are a few places where due to the routing, the object type gets cut off. For example, I have an endpoint for `employees/{id}` and I don't see the `employees` part. What I'm trying to do is ensure that I'm using the correct ID, so I'd like to be able to say CreateId<T>(string id) and if the first part isn't the pluralized T to tack it on there, but I want to use the same underlying logic Raven does.

I'm in .Net 6 so I don't have the old framework pluralization I used to use (and IIRC, so did Raven)

**Comment by ayende** [comment]
To answer your question, we use: 
https://github.com/ravendb/ravendb/blob/d426f33da4b417d711da27b9f28e4889439a3773/src/Raven.Client/Documents/Conventions/DocumentConventions.cs#L909

However, that isn't what you want to do, you can use:

```
public string CreateId<T>(string id)
{
   var expected = DocumentStore.Conventions.GetCollectionName(typeof(T));
   if(id.StartsWith(expected) == false)
    return expected +"/" + id;
  return id;
}
```

Rough code, but should get the point across, I hope.

**Reply by ItsRobbAllen** [reply]
Perfect! Thank you.

**Comment by georgiosd** [comment]
@ItsRobbAllen FWIW, we get around this with the concept of "partial ids"

```
public static string KeyFromPartial<T>(this IDocumentStore store, object partialId)
        {
            var conventions = store.Conventions;

            string collectionName = conventions.GetCollectionName(typeof(T));
            string idPrefix = conventions.TransformTypeCollectionNameToDocumentIdPrefix(collectionName);

            if (partialId is not string s)
            {
                return $"{idPrefix}/{partialId}";
            }

            string normalized = Regex.Replace(s, @"\-[a-e]", match => match.Value.ToUpperInvariant());
                
            return $"{idPrefix}/{normalized}";
        }
```
If you're wondering about the regex, it's to uppercase the node letter so that direct id comparisons don't fail.

The controller code looks like this:

```
[Get("users/{partialId}"]
public async Task<IActionResult> Get(string partialId)
{
  string userId = session.KeyFromPartial<User>(partialId);
  ...
}
```
